spmd/compiler/fusion.py

```diff
@@ -162,22 +214,448 @@ def _scan_graph_for_fusion_elements(
+    in_dtype: Optional[torch.dtype] = None,
+    in_memory_format: Optional[torch.memory_format] = None,
+) -> TensorMetadata:
+    """update a node's metadata to the the new shape, dtype and/or memory format"""
+    curr = node.meta.get("tensor_meta")
+    assert (
+        curr is not None
+    ), f"failed to obtain tensor meta data on node {node.name}"
+
+    shape = curr.shape
+    curr_dtype = curr.dtype
+    requires_grad = curr.requires_grad
+    stride = curr.stride
+
+    curr_memory_format = curr.memory_format
+    is_quantized = curr.is_quantized
+    qparams = curr.qparams
+
+    updated_dtype = in_dtype if in_dtype is not None else curr_dtype
+    updated_memory_format = (
+        in_memory_format if in_memory_format is not None else curr_memory_format
+    )
+
+    # tempt = torch.empty(new_shape)
+    # new_shape = tempt.shape
+
+    new_metadata = TensorMetadata(
+        new_shape,
+        updated_dtype,
+        requires_grad,
+        stride,
+        updated_memory_format,
+        is_quantized,
+        qparams,
+    )
+
+    # update meta with new TensorMetadata
+    saved_meta = node.meta.get("tensor_meta")
+    node.meta["tensor_meta"] = new_metadata
+
+    return new_metadata
+
+
+def _finalize_output_node(
+    gi: GraphInfo,
+    gm: fx.GraphModule,
+    fe_list: List[FusionElement],
+    start: int,
+    stop: int,
+    new_output_args: List[fx.Node],
+) -> None:
+    """reworks output node args to original grad tensors, replacing the wait_comms
+    we update a copy of the output args, then finalized after all fusion is done."""
+
+    # output_node = gi.output
+    # new_output_args = []
+
+    # curr_output_args = list(gi.output.args[0])
+    replacement_mapping: Dict[fx.Node, fx.Node] = {}
+
+    # map out all updated nodes in our list
+    for item in fe_list:
+        grad_node = item.grad_tensor_node
+        wait_node = item.wait_node
+        replacement_mapping[wait_node] = grad_node  # type: ignore
+
+    # we have fused a subset, only update that subset within the larger output node args
+    # TODO - this assumes that all gradient tensors are comm handled.
+    for i in range(len(fe_list)):
+        index = start + i
+        curr_node = new_output_args[index]
+
+        if curr_node is not None:
+            assert curr_node.name.startswith(
+                "wait"
+            ), f"Non comm gradient output tensor incorrectly handled...needs fix. {new_output_args[start+i]}"
+            new_output_args[start + i] = replacement_mapping[curr_node]
+
+    # _debug(f"537 - updated output args = {new_output_args}\n")
+
+
+def _determine_peak_memory(gi: GraphInfo, fusion_policy: int) -> int:
+    """
+    scans fe list to determine max memory required across all fusion instances.
+    this result is used to allocate the global buffer for fusion, where we
+    re-use a global buffer to avoid repeated allocations per fusion.
+    """
+    peak_memory = 0  # currently measured in numel
```

Comment:
This will depend on the precision used for grad reduction right? Should we figure out a way to incorporate that policy here?

Reply:
yes correct.  I've entered a tracking issue so this can be added after the PR.  
https://github.com/pytorch/tau/issues/656